<h1>Model Training</h1>

<h3>Import Data and Required Packages</h3>
<p>Import PANDAS, NUMPY, MATPLOTLIB, SEABORN, WARNINGS</p>



In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

<h3>Import CSV Data as Pandas DataFrame</h3>

In [4]:
df = pd.read_csv('data/student_depression_dataset.csv')

***Showing Top 5 and Bottom 5 ROWS***

In [5]:
df.head()

id  Gender   Age           City Profession  Academic Pressure  \
0   2    Male  33.0  Visakhapatnam    Student                5.0   
1   8  Female  24.0      Bangalore    Student                2.0   
2  26    Male  31.0       Srinagar    Student                3.0   
3  30  Female  28.0       Varanasi    Student                3.0   
4  32  Female  25.0         Jaipur    Student                4.0   

   Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
0            0.0  8.97                 2.0               0.0   
1            0.0  5.90                 5.0               0.0   
2            0.0  7.03                 5.0               0.0   
3            0.0  5.59                 2.0               0.0   
4            0.0  8.13                 3.0               0.0   

        Sleep Duration Dietary Habits   Degree  \
0          '5-6 hours'        Healthy  B.Pharm   
1          '5-6 hours'       Moderate      BSc   
2  'Less than 5 hours'        Healthy       BA   
3          '7-8 hours'       Moderate      BCA   
4          '5-6 hours'       Moderate   M.Tech   

  Have you ever had suicidal thoughts ?  Work/Study Hours Financial Stress  \
0                                   Yes               3.0              1.0   
1                                    No               3.0              2.0   
2                                    No               9.0              1.0   
3                                   Yes               4.0              5.0   
4                                   Yes               1.0              1.0   

  Family History of Mental Illness  Depression  
0                               No           1  
1                              Yes           0  
2                              Yes           0  
3                              Yes           1  
4                               No           0

In [6]:
df.tail()

id  Gender   Age       City Profession  Academic Pressure  \
27896  140685  Female  27.0      Surat    Student                5.0   
27897  140686    Male  27.0   Ludhiana    Student                2.0   
27898  140689    Male  31.0  Faridabad    Student                3.0   
27899  140690  Female  18.0   Ludhiana    Student                5.0   
27900  140699    Male  27.0      Patna    Student                4.0   

       Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
27896            0.0  5.75                 5.0               0.0   
27897            0.0  9.40                 3.0               0.0   
27898            0.0  6.61                 4.0               0.0   
27899            0.0  6.88                 2.0               0.0   
27900            0.0  9.24                 1.0               0.0   

            Sleep Duration Dietary Habits      Degree  \
27896          '5-6 hours'      Unhealthy  'Class 12'   
27897  'Less than 5 hours'        Healthy         MSc   
27898          '5-6 hours'      Unhealthy          MD   
27899  'Less than 5 hours'        Healthy  'Class 12'   
27900  'Less than 5 hours'        Healthy         BCA   

      Have you ever had suicidal thoughts ?  Work/Study Hours  \
27896                                   Yes               7.0   
27897                                    No               0.0   
27898                                    No              12.0   
27899                                   Yes              10.0   
27900                                   Yes               2.0   

      Financial Stress Family History of Mental Illness  Depression  
27896              1.0                              Yes           0  
27897              3.0                              Yes           0  
27898              2.0                               No           0  
27899              5.0                               No           1  
27900              3.0                              Yes           1

***Information About the Dataset***

In [7]:
df.describe()

id           Age  Academic Pressure  Work Pressure  \
count   27901.000000  27901.000000       27901.000000   27901.000000   
mean    70442.149421     25.822300           3.141214       0.000430   
std     40641.175216      4.905687           1.381465       0.043992   
min         2.000000     18.000000           0.000000       0.000000   
25%     35039.000000     21.000000           2.000000       0.000000   
50%     70684.000000     25.000000           3.000000       0.000000   
75%    105818.000000     30.000000           4.000000       0.000000   
max    140699.000000     59.000000           5.000000       5.000000   

               CGPA  Study Satisfaction  Job Satisfaction  Work/Study Hours  \
count  27901.000000        27901.000000      27901.000000      27901.000000   
mean       7.656104            2.943837          0.000681          7.156984   
std        1.470707            1.361148          0.044394          3.707642   
min        0.000000            0.000000          0.000000          0.000000   
25%        6.290000            2.000000          0.000000          4.000000   
50%        7.770000            3.000000          0.000000          8.000000   
75%        8.920000            4.000000          0.000000         10.000000   
max       10.000000            5.000000          4.000000         12.000000   

         Depression  
count  27901.000000  
mean       0.585499  
std        0.492645  
min        0.000000  
25%        0.000000  
50%        1.000000  
75%        1.000000  
max        1.000000

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27901 entries, 0 to 27900
Data columns (total 18 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   id                                     27901 non-null  int64  
 1   Gender                                 27901 non-null  object 
 2   Age                                    27901 non-null  float64
 3   City                                   27901 non-null  object 
 4   Profession                             27901 non-null  object 
 5   Academic Pressure                      27901 non-null  float64
 6   Work Pressure                          27901 non-null  float64
 7   CGPA                                   27901 non-null  float64
 8   Study Satisfaction                     27901 non-null  float64
 9   Job Satisfaction                       27901 non-null  float64
 10  Sleep Duration                         27901 non-null  object 
 11  Di

In [9]:
df.shape

(27901, 18)

In [10]:
df.dtypes

id                                         int64
Gender                                    object
Age                                      float64
City                                      object
Profession                                object
Academic Pressure                        float64
Work Pressure                            float64
CGPA                                     float64
Study Satisfaction                       float64
Job Satisfaction                         float64
Sleep Duration                            object
Dietary Habits                            object
Degree                                    object
Have you ever had suicidal thoughts ?     object
Work/Study Hours                         float64
Financial Stress                          object
Family History of Mental Illness          object
Depression                                 int64
dtype: object

In [11]:
# Dropped the ID and CITY columns
df.drop(columns=['id', 'City'], inplace=True, axis=1)

In [12]:
columns = df.columns
columns

Index(['Gender', 'Age', 'Profession', 'Academic Pressure', 'Work Pressure',
       'CGPA', 'Study Satisfaction', 'Job Satisfaction', 'Sleep Duration',
       'Dietary Habits', 'Degree', 'Have you ever had suicidal thoughts ?',
       'Work/Study Hours', 'Financial Stress',
       'Family History of Mental Illness', 'Depression'],
      dtype='object')

In [13]:
df[columns].nunique()

Gender                                     2
Age                                       34
Profession                                14
Academic Pressure                          6
Work Pressure                              3
CGPA                                     332
Study Satisfaction                         6
Job Satisfaction                           5
Sleep Duration                             5
Dietary Habits                             4
Degree                                    28
Have you ever had suicidal thoughts ?      2
Work/Study Hours                          13
Financial Stress                           6
Family History of Mental Illness           2
Depression                                 2
dtype: int64

In [14]:
for column in columns:
    print(column)
    print(df[column].unique())

Gender
['Male' 'Female']
Age
[33. 24. 31. 28. 25. 29. 30. 27. 19. 20. 23. 18. 21. 22. 34. 32. 26. 39.
 35. 42. 36. 58. 49. 38. 51. 44. 43. 46. 59. 54. 48. 56. 37. 41.]
Profession
['Student' "'Civil Engineer'" 'Architect' "'UX/UI Designer'"
 "'Digital Marketer'" "'Content Writer'" "'Educational Consultant'"
 'Teacher' 'Manager' 'Chef' 'Doctor' 'Lawyer' 'Entrepreneur' 'Pharmacist']
Academic Pressure
[5. 2. 3. 4. 1. 0.]
Work Pressure
[0. 5. 2.]
CGPA
[ 8.97    5.9     7.03    5.59    8.13    5.7     9.54    8.04    9.79
  8.38    6.1     7.04    8.52    5.64    8.58    6.51    7.25    7.83
  9.93    8.74    6.73    5.57    8.59    7.1     6.08    5.74    9.86
  6.7     6.21    5.87    6.37    9.72    5.88    9.56    6.99    5.24
  9.21    7.85    6.95    5.86    7.92    9.66    8.94    9.71    7.87
  5.6     7.9     5.46    6.79    8.7     7.38    8.5     7.09    9.82
  8.89    7.94    9.11    6.75    7.53    9.49    9.01    7.64    5.27
  6.      9.44    5.75    7.51    9.05    6.38    8.

***Checking and NULL And Missing Values and Filling them***

In [15]:
df.isnull()

Gender    Age  Profession  Academic Pressure  Work Pressure   CGPA  \
0       False  False       False              False          False  False   
1       False  False       False              False          False  False   
2       False  False       False              False          False  False   
3       False  False       False              False          False  False   
4       False  False       False              False          False  False   
...       ...    ...         ...                ...            ...    ...   
27896   False  False       False              False          False  False   
27897   False  False       False              False          False  False   
27898   False  False       False              False          False  False   
27899   False  False       False              False          False  False   
27900   False  False       False              False          False  False   

       Study Satisfaction  Job Satisfaction  Sleep Duration  Dietary Habits  \
0                   False             False           False           False   
1                   False             False           False           False   
2                   False             False           False           False   
3                   False             False           False           False   
4                   False             False           False           False   
...                   ...               ...             ...             ...   
27896               False             False           False           False   
27897               False             False           False           False   
27898               False             False           False           False   
27899               False             False           False           False   
27900               False             False           False           False   

       Degree  Have you ever had suicidal thoughts ?  Work/Study Hours  \
0       False                                  False             False   
1       False                                  False             False   
2       False                                  False             False   
3       False                                  False             False   
4       False                                  False             False   
...       ...                                    ...               ...   
27896   False                                  False             False   
27897   False                                  False             False   
27898   False                                  False             False   
27899   False                                  False             False   
27900   False                                  False             False   

       Financial Stress  Family History of Mental Illness  Depression  
0                 False                             False       False  
1                 False                             False       False  
2                 False                             False       False  
3                 False                             False       False  
4                 False                             False       False  
...                 ...                               ...         ...  
27896             False                             False       False  
27897             False                             False       False  
27898             False                             False       False  
27899             False                             False       False  
27900             False                             False       False  

[27901 rows x 16 columns]

In [16]:
df.isnull().sum()

Gender                                   0
Age                                      0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         0
Family History of Mental Illness         0
Depression                               0
dtype: int64

In [17]:
df.isna().any()

Gender                                   False
Age                                      False
Profession                               False
Academic Pressure                        False
Work Pressure                            False
CGPA                                     False
Study Satisfaction                       False
Job Satisfaction                         False
Sleep Duration                           False
Dietary Habits                           False
Degree                                   False
Have you ever had suicidal thoughts ?    False
Work/Study Hours                         False
Financial Stress                         False
Family History of Mental Illness         False
Depression                               False
dtype: bool

In [18]:
(df == " ").sum()

Gender                                   0
Age                                      0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         0
Family History of Mental Illness         0
Depression                               0
dtype: int64

In [19]:
df.isin(['NA', 'NULL', '-', '?']).sum()

Gender                                   0
Age                                      0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         3
Family History of Mental Illness         0
Depression                               0
dtype: int64

<p>I found 3 missing values with "?" in Financial Stress Column</p>

In [20]:
# Now Handling the Missing values

# Replacing the ? value with the NaN
df['Financial Stress'].replace('?', np.nan, inplace=True)

# Filling the NaN value with the MODE Imputer coz, its a categorical Column
df['Financial Stress'].fillna(df['Financial Stress'].mode()[0], inplace=True)

# Converting the Categorical column to Numerical Column
df['Financial Stress']=df['Financial Stress'].astype('float')


In [21]:
df.isin(['NA', 'NULL', '-', '?']).sum()

Gender                                   0
Age                                      0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         0
Family History of Mental Illness         0
Depression                               0
dtype: int64

In [22]:
print(df['Financial Stress'].unique())
print(df['Financial Stress'].value_counts())

[1. 2. 5. 3. 4.]
Financial Stress
5.0    6718
4.0    5775
3.0    5226
1.0    5121
2.0    5061
Name: count, dtype: int64


***Retriving the Numerical Columns***

In [23]:
numerical_columns = df.select_dtypes(include=['number']).columns.tolist()
numerical_columns

['Age',
 'Academic Pressure',
 'Work Pressure',
 'CGPA',
 'Study Satisfaction',
 'Job Satisfaction',
 'Work/Study Hours',
 'Financial Stress',
 'Depression']

In [24]:
categorical_columns = df.select_dtypes(include=['object']).columns.tolist()
categorical_columns

['Gender',
 'Profession',
 'Sleep Duration',
 'Dietary Habits',
 'Degree',
 'Have you ever had suicidal thoughts ?',
 'Family History of Mental Illness']

In [25]:
for column in categorical_columns:
    print(column)
    print(df[column].unique())

Gender
['Male' 'Female']
Profession
['Student' "'Civil Engineer'" 'Architect' "'UX/UI Designer'"
 "'Digital Marketer'" "'Content Writer'" "'Educational Consultant'"
 'Teacher' 'Manager' 'Chef' 'Doctor' 'Lawyer' 'Entrepreneur' 'Pharmacist']
Sleep Duration
["'5-6 hours'" "'Less than 5 hours'" "'7-8 hours'" "'More than 8 hours'"
 'Others']
Dietary Habits
['Healthy' 'Moderate' 'Unhealthy' 'Others']
Degree
['B.Pharm' 'BSc' 'BA' 'BCA' 'M.Tech' 'PhD' "'Class 12'" 'B.Ed' 'LLB' 'BE'
 'M.Ed' 'MSc' 'BHM' 'M.Pharm' 'MCA' 'MA' 'B.Com' 'MD' 'MBA' 'MBBS' 'M.Com'
 'B.Arch' 'LLM' 'B.Tech' 'BBA' 'ME' 'MHM' 'Others']
Have you ever had suicidal thoughts ?
['Yes' 'No']
Family History of Mental Illness
['No' 'Yes']


In [26]:
df.shape

(27901, 16)

<h3>Preparing X varible and Y variable</h3>

In [27]:
X = df.drop(columns=['Depression'], axis=1)

In [28]:
X.head()

Gender   Age Profession  Academic Pressure  Work Pressure  CGPA  \
0    Male  33.0    Student                5.0            0.0  8.97   
1  Female  24.0    Student                2.0            0.0  5.90   
2    Male  31.0    Student                3.0            0.0  7.03   
3  Female  28.0    Student                3.0            0.0  5.59   
4  Female  25.0    Student                4.0            0.0  8.13   

   Study Satisfaction  Job Satisfaction       Sleep Duration Dietary Habits  \
0                 2.0               0.0          '5-6 hours'        Healthy   
1                 5.0               0.0          '5-6 hours'       Moderate   
2                 5.0               0.0  'Less than 5 hours'        Healthy   
3                 2.0               0.0          '7-8 hours'       Moderate   
4                 3.0               0.0          '5-6 hours'       Moderate   

    Degree Have you ever had suicidal thoughts ?  Work/Study Hours  \
0  B.Pharm                                   Yes               3.0   
1      BSc                                    No               3.0   
2       BA                                    No               9.0   
3      BCA                                   Yes               4.0   
4   M.Tech                                   Yes               1.0   

   Financial Stress Family History of Mental Illness  
0               1.0                               No  
1               2.0                              Yes  
2               1.0                              Yes  
3               5.0                              Yes  
4               1.0                               No

In [29]:
X.shape

(27901, 15)

In [30]:
y = df['Depression']

In [31]:
y

0        1
1        0
2        0
3        1
4        0
        ..
27896    0
27897    0
27898    0
27899    1
27900    1
Name: Depression, Length: 27901, dtype: int64

<h3>Encoding and Feature Engineering</h3>
<p><b>1. Encoding</b></p>
<table>
  <tr>
    <th>Column Name</th>
    <th>Encoding Method</th>
    <th>Reason</th>
  </tr>
  <tr>
    <td>Gender</td>
    <td>One-Hot Encoding ✅</td>
    <td>Only 2 categories (Male, Female)</td>
  </tr>
  <tr>
    <td>Profession</td>
    <td>Target Encoding (or One-Hot if few unique values) ✅</td>
    <td>Many unique values, possible ranking</td>
  </tr>
  <tr>
    <td>Sleep Duration</td>
    <td>Label Encoding ✅</td>
    <td>Ordered categories (Less than 5 hours < 5-6 hours < 7-8 hours)</td>
  </tr>
  <tr>
    <td>Dietary Habits</td>
    <td>Label Encoding ✅</td>
    <td>Ordered categories (Healthy > Moderate > Unhealthy)</td>
  </tr>
  <tr>
    <td>Degree</td>
    <td>Target Encoding ✅</td>
    <td>Many unique values, possible ranking</td>
  </tr>
  <tr>
    <td>Suicidal Thoughts?</td>
    <td>Label Encoding ✅</td>
    <td>Binary (Yes = 1, No = 0)</td>
  </tr>
  <tr>
    <td>Family History of Mental Illness</td>
    <td>Label Encoding ✅</td>
    <td>Binary (Yes = 1, No = 0)</td>
  </tr>
</table>

<p><b>2. Scaling Numerical Features</b></p>

In [32]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
import category_encoders as ce 

# Define Categorical and numerical Features
num_features = X.select_dtypes(include=['number']).columns
cat_nominal_features = ['Gender']
cat_ordinal_features = ["Sleep Duration", "Dietary Habits"]
cat_target_encoding_features = ['Degree', 'Profession']
binary_features = ["Have you ever had suicidal thoughts ?", "Family History of Mental Illness"]

# splitting the dataset 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

one_hot_encoder = OneHotEncoder(handle_unknown='ignore')
numeric_transformer = StandardScaler()

# Label Encoding (Ordinal and Binary)
label_encoders = {}
for col in cat_ordinal_features + binary_features:
    le = LabelEncoder()
    X_train[col] = le.fit_transform(X_train[col])
    X_test[col]  = le.transform(X_test[col])
    label_encoders[col] = le

# Target Encoding
target_encoder = ce.TargetEncoder(cols = cat_target_encoding_features)
X_train[cat_target_encoding_features] = target_encoder.fit_transform(X_train[cat_target_encoding_features], y_train)
X_test[cat_target_encoding_features]  = target_encoder.transform(X_test[cat_target_encoding_features])

# Define Column Transformer
preprocessor = ColumnTransformer(
    [
        ('OneHotEncoder', one_hot_encoder, cat_nominal_features),
        ('StandardScaler', numeric_transformer, num_features),
        ('StandardScaler_ordinal_target', numeric_transformer, cat_ordinal_features + cat_target_encoding_features)
    ],
    remainder='passthrough'
)

# Apply Transformation
X_train_transformed = preprocessor.fit_transform(X_train)
X_test_transformed  = preprocessor.transform(X_test)


In [33]:
X_train.shape, X_test.shape

((22320, 15), (5581, 15))

In [34]:
X_train_transformed.shape, X_test_transformed.shape

((22320, 16), (5581, 16))

<h2>Machine Learning Models</h2>

In [35]:
## Train a Decision Tree Model 
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, recall_score

# Initialize and Train a model
dt_model = DecisionTreeClassifier(random_state = 42)
dt_model.fit(X_train_transformed, y_train)

# Make Predictions
y_pred_dt = dt_model.predict(X_test_transformed)

# Evaluate the model
accuracy_dt = accuracy_score(y_test, y_pred_dt)
recall_dt = recall_score(y_test, y_pred_dt)
print(f" Decision Tree Model Accuracy : {accuracy_dt*100:.2f}")
print(f" Decision Tree Model Recall : {recall_dt*100:.2f}")


 Decision Tree Model Accuracy : 75.18
 Decision Tree Model Recall : 78.01


In [36]:
## Train Logistic Regression Model
from sklearn.linear_model import LogisticRegression

# Initialize and Train a Model
lr_model = LogisticRegression()
lr_model.fit(X_train_transformed, y_train)

# Make Prediction
y_pred_lr = lr_model.predict(X_test_transformed)

# Evaluate the Model
accuracy_lr = accuracy_score(y_test, y_pred_lr)
recall_lr = recall_score(y_test, y_pred_lr)
print(f"Logistic Regression Model Accuracy : {accuracy_lr*100:.2f}")
print(f" Logistic Regression Model Recall : {recall_lr*100:.2f}")


Logistic Regression Model Accuracy : 83.69
 Logistic Regression Model Recall : 87.25


In [37]:
## Train Random Forest Model
from sklearn.ensemble import RandomForestClassifier

# Initialize and Train a model
rfc_model = RandomForestClassifier(n_estimators = 100, random_state = 42)
rfc_model.fit(X_train_transformed, y_train)

# Make Prediction
y_pred_rfc = rfc_model.predict(X_test_transformed)

# Evaluate the model
accuracy_rfc = accuracy_score(y_test, y_pred_rfc)
recall_rfc = recall_score(y_test, y_pred_rfc)
print(f"Random Forest Classifier Model Accuracy : {accuracy_rfc*100:.2f}")
print(f"Random Forest Classifier Model Recall : {recall_rfc*100:.2f}")


Random Forest Classifier Model Accuracy : 82.92
Random Forest Classifier Model Recall : 86.29


In [38]:
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from xgboost import XGBClassifier
from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier
from catboost import CatBoostClassifier

# Define Classifiers
model = {
    'SVM' : SVC(),
    'KNN' : KNeighborsClassifier(),
    'XGBoost' : XGBClassifier(),
    'CatBoost' : CatBoostClassifier(verbose = False),
    'AdaBoost' : AdaBoostClassifier(),
    'Gradient Boosting' : GradientBoostingClassifier()
}

# Train and Evaluate
for name, model in model.items():
    model.fit(X_train_transformed, y_train)
    y_pred = model.predict(X_test_transformed)
    acc = accuracy_score(y_test, y_pred)
    recall_all = recall_score(y_test, y_pred)
    print(f"{name} Accuracy : {acc*100:.2f}")
    print(f"{name} Recall Score -> {recall_all*100:.2f}")

SVM Accuracy : 83.30
SVM Recall Score -> 87.77
KNN Accuracy : 80.83
KNN Recall Score -> 87.06
XGBoost Accuracy : 83.07
XGBoost Recall Score -> 86.57
CatBoost Accuracy : 83.50
CatBoost Recall Score -> 87.18
AdaBoost Accuracy : 84.00
AdaBoost Recall Score -> 87.71
Gradient Boosting Accuracy : 83.68
Gradient Boosting Recall Score -> 87.62


### ✅ Primary Metric: Recall

**In medical or mental health classification tasks** (like *depression detection*), **Recall** is the most important metric for the positive class (i.e., identifying depressed students).

#### Why?

- **Recall** (Sensitivity) ensures that *most actual depressed students* are correctly identified.
- **False negatives** (missed cases) can be dangerous — a depressed student missed by the model might not get the help they need.

---

### ✅ Best Approach: Use Both Recall & Accuracy Together

While **Recall** is critical in sensitive applications, combining it with **Accuracy** gives a more balanced view of overall model performance.


<h2>Classification Model Performance (Before Fine-Tuning)</h2>

<table>
  <thead>
    <tr>
      <th>Model</th>
      <th>Accuracy (%)</th>
      <th>Recall (%)</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>SVM</td><td>83.30</td><td>87.77</td></tr>
    <tr><td>KNN</td><td>80.83</td><td>87.06</td></tr>
    <tr><td>XGBoost</td><td>83.07</td><td>86.57</td></tr>
    <tr><td>CatBoost</td><td>83.50</td><td>87.18</td></tr>
    <tr><td>AdaBoost</td><td>84.00</td><td>87.71</td></tr>
    <tr><td>Gradient Boosting</td><td>83.69</td><td>87.62</td></tr>
    <tr><td>Random Forest</td><td>82.92</td><td>86.29</td></tr>
    <tr><td>Logistic Regression</td><td>83.69</td><td>87.25</td></tr>
    <tr><td>Decision Tree</td><td>75.18</td><td>78.01</td></tr>
  </tbody>
</table>

<h2>Insights</h2>
<p><strong>Top Models Based on Recall & Accuracy:</strong></p>
<ul>
  <li><strong>AdaBoost</strong> – Highest accuracy (84.00%) and very high recall (87.71%)</li>
  <li><strong>Gradient Boosting</strong> – Strong performance with 83.69% accuracy and 87.62% recall</li>
  <li><strong>Support Vector Machine (SVM)</strong> – Recall: 87.77%, Accuracy: 83.30%</li>
  <li><strong>Logistic Regression</strong> – Accuracy: 83.69%, Recall: 87.25%</li>
  <li><strong>CatBoost</strong> – Accuracy: 83.50%, Recall: 87.18%</li>
</ul>

<p><strong>Special Mention:</strong></p>
<ul>
  <li><strong>K-Nearest Neighbors (KNN)</strong> – Although its accuracy is slightly lower (80.83%), its recall is high (87.06%), making it valuable in medical classification scenarios where recall is critical.</li>
</ul>

<h2>Next Steps: Fine-Tuning</h2>
<p>To maximize performance, consider hyperparameter tuning on the top-performing models, especially:</p>
<ul>
  <li><strong>AdaBoost</strong> (best overall trade-off)</li>
  <li><strong>Gradient Boosting</strong> (consistent recall/accuracy)</li>
  <li><strong>Logistic Regression</strong> (simple and high-performing)</li>
  <li><strong>KNN</strong> (to try boosting accuracy without compromising recall)</li>
</ul>

<p><strong>Best Approach:</strong> Use both <strong>Recall</strong> and <strong>Accuracy</strong> together to evaluate the model's robustness. In mental health applications like depression detection, <strong>Recall</strong> is the primary metric to ensure no true cases are missed.</p>


In [39]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report

# Assuming X_train, y_train, X_test, y_test are already defined and preprocessed

# Dictionary of classifiers (you can replace with your own model instances)
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB

models = {
    "Logistic Regression": LogisticRegression(),
    "Random Forest": RandomForestClassifier(),
    "Decision Tree": DecisionTreeClassifier(),
    "Support Vector Machine": SVC(),
    "K-Nearest Neighbors": KNeighborsClassifier(),
    "Naive Bayes": GaussianNB(),
    "AdaBoost": AdaBoostClassifier(),
    "Gradient Boosting": GradientBoostingClassifier()
}

# Your hyperparameter grid
params = {
    "Logistic Regression": {
        'C': [0.01, 0.1, 1, 10, 100],
        'penalty': ['l2'],
        'solver': ['liblinear', 'saga'],
        'class_weight': [None, 'balanced'],
        'max_iter': [1000]
    },
    "Random Forest": {
        'n_estimators': [100, 200],
        'max_depth': [None, 10, 20],
        'min_samples_split': [2, 5],
        'min_samples_leaf': [1, 2],
        'class_weight': [None, 'balanced']
    },
    "Decision Tree": {
        'criterion': ['gini', 'entropy'],
        'max_depth': [None, 10, 20],
        'min_samples_split': [2, 5],
        'min_samples_leaf': [1, 2],
        'class_weight': [None, 'balanced']
    },
    "Support Vector Machine": {
        'C': [0.1, 1, 10],
        'kernel': ['linear', 'rbf'],
        'gamma': ['scale', 'auto'],
        'class_weight': [None, 'balanced']
    },
    "K-Nearest Neighbors": {
        'n_neighbors': [3, 5, 7, 11],
        'weights': ['uniform', 'distance'],
        'metric': ['euclidean', 'manhattan']
    },
    "Naive Bayes": {},
    "AdaBoost": {
        'n_estimators': [100, 200],
        'learning_rate': [0.01, 0.1, 1]
    },
    "Gradient Boosting": {
        'n_estimators': [100, 200],
        'learning_rate': [0.01, 0.1],
        'subsample': [0.8, 1.0],
        'max_depth': [3, 5]
    }
}

best_models = {}

for name, model in models.items():
    print(f"\n\033[94mTuning {name}...\033[0m")
    param_grid = params[name]
    if param_grid:
        grid = GridSearchCV(model, param_grid, cv=5, scoring='recall', n_jobs=-1)
        grid.fit(X_train_transformed, y_train)
        best_model = grid.best_estimator_
        print(f"Best Params: {grid.best_params_}")
    else:
        model.fit(X_train_transformed, y_train)
        best_model = model
    
    best_models[name] = best_model
    y_pred = best_model.predict(X_test_transformed)
    accuracy_scores = accuracy_score(y_test, y_pred)
    recall_scores = recall_score(y_test, y_pred)
    print(classification_report(y_test, y_pred))
    print(f'{name} Model Accuracy Score : {accuracy_scores*100:.5f}')
    print(f'{name} Model Recall Score : {recall_scores*100:.5f}')

# You can now compare best_models to select top-performing models.



Tuning Logistic Regression...
Best Params: {'C': 0.01, 'class_weight': None, 'max_iter': 1000, 'penalty': 'l2', 'solver': 'liblinear'}
              precision    recall  f1-score   support

           0       0.82      0.78      0.80      2343
           1       0.85      0.88      0.86      3238

    accuracy                           0.84      5581
   macro avg       0.83      0.83      0.83      5581
weighted avg       0.84      0.84      0.84      5581

Logistic Regression Model Accuracy Score : 83.62301
Logistic Regression Model Recall Score : 87.67758

Tuning Random Forest...
Best Params: {'class_weight': None, 'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}
              precision    recall  f1-score   support

           0       0.82      0.77      0.79      2343
           1       0.84      0.87      0.86      3238

    accuracy                           0.83      5581
   macro avg       0.83      0.82      0.83      5581
weighted avg     

<h2>✅ Primary Metric: Recall</h2>

<p><strong>In medical or mental health classification tasks</strong> (like <em>depression detection</em>), 
<strong>Recall</strong> is the most important metric for the positive class (i.e., identifying depressed students).</p>

<h3>Why?</h3>
<ul>
  <li><strong>Recall</strong> (Sensitivity) ensures that <em>most actual depressed students</em> are correctly identified.</li>
  <li><strong>False negatives</strong> (missed cases) can be dangerous — a depressed student missed by the model might not get the help they need.</li>
</ul>

<hr>

<h2>✅ Best Approach: Use Both Recall & Accuracy Together</h2>

<p>While <strong>Recall</strong> is critical in sensitive applications, combining it with <strong>Accuracy</strong> gives a more balanced view of overall model performance.</p>

<hr>

<h2>Classification Model Performance (After Fine-Tuning)</h2>

<table>
  <thead>
    <tr>
      <th>Model</th>
      <th>Accuracy (%)</th>
      <th>Recall (%)</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>AdaBoost</td><td>82.92</td><td>89.22</td></tr>
    <tr><td>Gradient Boosting</td><td>82.73</td><td>89.68</td></tr>
    <tr><td>K-Nearest Neighbors</td><td>81.90</td><td>88.39</td></tr>
    <tr><td>Support Vector Machine</td><td>83.59</td><td>87.80</td></tr>
    <tr><td>Logistic Regression</td><td>83.62</td><td>87.68</td></tr>
    <tr><td>Random Forest</td><td>83.26</td><td>87.37</td></tr>
    <tr><td>Decision Tree</td><td>81.47</td><td>85.33</td></tr>
    <tr><td>Naive Bayes</td><td>82.08</td><td>80.36</td></tr>
  </tbody>
</table>

<hr>

<h2>📌 Insights</h2>

<h3>Top Models Based on Recall & Accuracy:</h3>
<ul>
  <li><strong>Gradient Boosting</strong> – Highest recall (89.68%) with solid accuracy (82.73%)</li>
  <li><strong>AdaBoost</strong> – High recall (89.22%) and decent accuracy (82.92%)</li>
  <li><strong>K-Nearest Neighbors</strong> – High recall (88.39%) with slight accuracy tradeoff (81.90%)</li>
  <li><strong>Support Vector Machine (SVM)</strong> – Recall: 87.80%, Accuracy: 83.59%</li>
  <li><strong>Logistic Regression</strong> – Simple, interpretable, Recall: 87.68%, Accuracy: 83.62%</li>
</ul>


In [40]:
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score, recall_score

best_models = {}

for name, model in models.items():
    print(f"\n\033[94mTuning {name}...\033[0m")
    param_grid = params[name]
    
    if param_grid:
        grid = GridSearchCV(model, param_grid, cv=5, scoring='recall', n_jobs=-1)
        grid.fit(X_train_transformed, y_train)
        best_model = grid.best_estimator_
        print(f"Best Params: {grid.best_params_}")
    else:
        model.fit(X_train_transformed, y_train)
        best_model = model
    
    best_models[name] = best_model
    
    y_pred = best_model.predict(X_test_transformed)

    # Evaluation metrics
    acc = accuracy_score(y_test, y_pred)
    rec = recall_score(y_test, y_pred)
    cm = confusion_matrix(y_test, y_pred)

    # Extract TP and FN
    tn, fp, fn, tp = cm.ravel()

    print("\nClassification Report:")
    print(classification_report(y_test, y_pred))
    
    print(f"{name} Model Accuracy Score : {acc*100:.5f}")
    print(f"{name} Model Recall Score   : {rec*100:.5f}")
    print("Confusion Matrix:")
    print(cm)
    print(f"True Positives (TP): {tp}")
    print(f"False Negatives (FN): {fn}")



Tuning Logistic Regression...
Best Params: {'C': 0.01, 'class_weight': None, 'max_iter': 1000, 'penalty': 'l2', 'solver': 'liblinear'}

Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.78      0.80      2343
           1       0.85      0.88      0.86      3238

    accuracy                           0.84      5581
   macro avg       0.83      0.83      0.83      5581
weighted avg       0.84      0.84      0.84      5581

Logistic Regression Model Accuracy Score : 83.62301
Logistic Regression Model Recall Score   : 87.67758
Confusion Matrix:
[[1828  515]
 [ 399 2839]]
True Positives (TP): 2839
False Negatives (FN): 399

Tuning Random Forest...
Best Params: {'class_weight': None, 'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 5, 'n_estimators': 100}

Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.78      0.80      2343
           1       0.85      0.87